Explicitly check if URI is file when checking for an absolute URI

Daniel15 opened the pull request:
Parsing `/foo.css` as a URI on Linux results in `file:///foo.css`, because `/foo.css` is a legitimate absolute path to a file on the file system. This code needs to explicitly check if the resulting URI is that of a file. 

After fixing this, all the disabled tests relating to this feature actually work on Mono, and can be re-enabled. This means that https://github.com/aspnet/External/issues/21 (which I can't see because it's private, but it was referenced in the code) can be closed as it's not an actual issue.

#2979

dnfclas commented:
Hi __@Daniel15__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

Daniel15 commented:
Actually I should be able to just use the existing test and remove the `FrameworkSkipCondition`, since that test should work on Mono now.

sornaks commented on test/Microsoft.AspNet.Mvc.TagHelpers.Test/ImageTagHelperTest.cs:
```diff
@@ -82,9 +82,6 @@ public void PreservesOrderOfSourceAttributesWhenRun()
             }
         }
 
-        [ConditionalTheory]
```
Since you removed the `[ConditionalTheory]` attribute, these tests are currently not running on any platforms. Replace
```c#
[ConditionalTheory]
// Mono issue - https://github.com/aspnet/External/issues/21
[FrameworkSkipCondition(RuntimeFrameworks.Mono)]
```
with
```c#
[Fact]
```
to make them run on all platforms.

Daniel15 replied:
Ohhhh, good catch! l didn't see any errors during the test run so I thought it was running :( I'll fix this tonight.

Daniel15 commented:
@sornaks I added the missing `[Fact]` attributes. Unfortunately I updated to the latest bits recently and my local environment seems pretty broken right now (Mono crashes with `* Assertion: should not be reached at class.c:6488` when I try to run these tests) so I'm going to see how the AppVeyor/TravisCI runs go.

sornaks commented:
Hmmm, your tests seem to have run in travis and is passing. There are a couple of other places where tests have been disabled with this issue (<a href="https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs#L63">here</a> and <a href="https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs#L128">here</a>). Can you enable them as well?

Daniel15 commented:
Those may need separate fixes, I haven't looked through much of the MVC code. I'd rather test and re-renable those other tests in a separate diff so as to not expand the scope of this one too much from its original intention (fixing the `asp-append-version` TagHelper). What do you think?

sornaks commented:
The places I mentioned are using `asp-append-version`. Ex: <a href="https://github.com/aspnet/Mvc/blob/dev/test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_Home/Script.cshtml#L120">Here</a>. I'm guessing, just removing the `if` statements in those 2 places shd have them working.

Daniel15 commented:
Looks like I'll need to rebase this on top of ca22f1649dca2ad0e4ba0643cf871184f526d644 - I'll do that and then try to enable the tests you mentioned @sornaks

dougbu commented:
@Daniel15 this PR is set. Let's :shipit: 

Please squash your commits (`git rebase -i head~3`) and confirm your fork is up-to-date. We'll get it in soon after.

Daniel15 commented:
Thanks @dougbu! Squashed it now, hopefully it worked right (first time I've ever squashed something in Git). I attempted to enable the tests in `test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs` too. The build apprears to be broken at the moment though -  I'm getting the same error that's currently on Travis for the dev branch:

```
/Users/danlo/src/AspNetMvc/src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs(8,27): DNX,Version=v4.5.1 error CS0234: The type or namespace name 'Primitives' does not exist in the namespace 'Microsoft.Framework' (are you missing an assembly reference?)
/Users/danlo/src/AspNetMvc/src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs(65,16): DNX,Version=v4.5.1 error CS0246: The type or namespace name 'StringValues' could not be found (are you missing a using directive or an assembly reference?)
/Users/danlo/src/AspNetMvc/src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs(41,36): DNX,Version=v4.5.1 error CS0246: The type or namespace name 'StringValues' could not be found (are you missing a using directive or an assembly reference?)
/Users/danlo/src/AspNetMvc/src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs(51,36): DNX,Version=v4.5.1 error CS0246: The type or namespace name 'StringValues' could not be found (are you missing a using directive or an assembly reference?)
/Users/danlo/src/AspNetMvc/src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs(35,74): DNX,Version=v4.5.1 error CS1503: Argument 1: cannot convert from 'string[]' to 'StringValues'
```

davidfowl commented:
@Daniel15 the build was fixed, could you rebase?

dougbu commented:
@davidfowl @Daniel15 is fine -- his fork was only a few commits behind and rebased cleanly for me. Will push shortly...

davidfowl commented:
@dougbu sweet

dougbu commented:
a2e53e8

Daniel15 commented:
Thank you :smile:
